elfi/methods/diagnostics.py

```diff
@@ -0,0 +1,294 @@
+    and the minimum-MRSSE combination is chosen as the one with the optimal performance.
+
+    Attributes
+    ----------
+    simulator : elfi.Node
+        Node (often elfi.Simulator) for which the summary statistics will be applied.
+        The node is the final node of a coherent ElfiModel (i.e. it has no child nodes).
+    fn_distance : str or callable function
+        Function for calculating distance (can be a string if the function is implemented in ELFI).
+    list_ss : List of callable functions, optional
+        List of uncombined candidate summary statistics.
+    prepared_ss : List of lists of callable functions, optional
+        List of combined candidate summary statistics.
+    max_cardinality : int, optional
+        Maximum cardinality of a candidate summary-statistics combination.
+    seed : int, optional
+
+    References
+    ----------
+    [1] Nunes, M. A., & Balding, D. J. (2010).
+    On optimal selection of summary statistics for approximate Bayesian computation.
+    Statistical applications in genetics and molecular biology, 9(1).
+    [2] Blum, M. G., Nunes, M. A., Prangle, D., & Sisson, S. A. (2013).
+    A comparative review of dimension reduction methods in approximate Bayesian computation.
+    Statistical Science, 28(2), 189-208.
+
+    # TODO: include different distance functions in the summary-statistics combinations.
+
+    """
+
+    def __init__(self, simulator, fn_distance, list_ss=None, prepared_ss=None,
+                 max_cardinality=4, seed=0):
+        """Initialise the summary-statistics selection for the Two Stage Procedure.
+
+        Parameters
+        ----------
+        simulator : elfi.Node
+            Node (often elfi.Simulator) for which the summary statistics will be applied.
+            The node is the final node of a coherent ElfiModel (i.e. it has no child nodes).
+        fn_distance : str or callable function
+            Distance metric (can be a string if the metric is implemented in ELFI).
+        list_ss : List of callable functions, optional
+            List of uncombined candidate summary statistics.
+        prepared_ss : List of lists of callable functions, optional
+            List of combined candidate summary statistics.
+        max_cardinality : int, optional
+            Maximum cardinality of a candidate summary-statistics combination.
+        seed : int, optional
+
+        """
+        if list_ss is None and prepared_ss is None:
+            raise ValueError('No summary statistics to assess.')
+
+        self.simulator = simulator
+        self.fn_distance = fn_distance
+        self.seed = seed
+        if prepared_ss is not None:
+            self.ss_candidates = prepared_ss
+        else:
+            self.ss_candidates = self._combine_ss(list_ss, max_cardinality=max_cardinality)
+        # Initialising an output pool as the rejection sampling will be used several times.
+        names_output = simulator.parents
```

Comment:
Sorry, I don't see a reason to save the parents in a general case (remember that the graph may be more complicated than just Priors and a Simulator). Ancestors possibly, but ELFI doesn't support it at the moment, although in this case it would of course be the same as all the nodes in the graph. 

But unless you use this data, it's just a waste of memory anyway. For reusing the simulations, it's enough to save the output from the simulator, as you do. One possible use for pooling the Priors would be to return indices instead of actual thetas from `_obtain_accepted_thetas`, but I think it's fine as it is also.

Reply:
Reverted the `OutputPool` to store the output from the simulator.